cuda/base/executor.cpp

```diff
@@ -223,12 +224,20 @@ void CudaExecutor::set_gpu_property()
 
 void CudaExecutor::init_handles()
 {
-    this->cublas_handle_ = handle_manager<cublasContext>(
-        kernels::cuda::cublas::init(), kernels::cuda::cublas::destroy);
-    typedef void (*func_ptr)(cusparseHandle_t);
-    this->cusparse_handle_ = handle_manager<cusparseContext>(
-        kernels::cuda::cusparse::init(),
-        static_cast<func_ptr>(kernels::cuda::cusparse::destroy));
+    if (device_id_ < this->get_num_devices() && device_id_ >= 0) {
+        const auto id = this->get_device_id();
+        device_guard g(id);
+        this->cublas_handle_ = handle_manager<cublasContext>(
+            kernels::cuda::cublas::init(), [id](cublasHandle_t handle) {
```

Comment:
What I don't understand is, if this is required for the `destroy` part, why not for the `init` one?

In addition, maybe we should pass the `id` as argument to the init/destroy and call the `device_guard` directly inside the respective functions, so that the base functions are now correct.

Reply:
The handle allocation is done in `init_handle`  by `kernels::cuda::cublas::init` so it is on the correct device due to `line 229: device_guard`.